### ISSUE TITLE
Automatic save and upload ftp bug

### ISSUE THREAD

**Issue by ElMarko34** [Natizyskunk/vscode-sftp#110]
**Do you read the FAQ?**
- [X] Yes.
- [ ] [I am going to read now.](https://github.com/Natizyskunk/vscode-sftp/blob/master/FAQ.md)

**Describe the bug**
**To Reproduce**
1) Create a file in minus characters exemple: interne.php
2) press ctrl+S to save and upload 
3) rename interne.php in Interne.php (note: first I is now in Caps)
4) press ctrl+S and look. File will be uploaded on name **interne**.php and not **Interne.php** (PROBLEM)
5) click right on Interne.php and upload it and no problem. Interne.php will be upload (NO PROBLEM)

The problemn is only when I press ctrl+S .
And the problem can be watched on console

[info] local ➞ remote c:\............\interne.php

[info] and [trace] talk about "interne.php" but this file doesn't exists because that was renamed

It look like there's a cache file that recorded file name in minus characters and if some characters become caps, ftp saver doesn't care.
Bug doesn't exist if we pres on upload file

**Comment by Natizyskunk** [comment]
Related issue : #24

**Comment by Natizyskunk** [comment]
Take a look at this configuration to make it work for now : https://github.com/Natizyskunk/vscode-sftp/issues/24#issuecomment-893860860

**Comment by microwin7** [comment]
> Take a look at [this configuration](https://github.com/Natizyskunk/vscode-sftp/issues/24#issuecomment-893860860) to make it work for now.
> 
> ```json
> "watcher": {
>     "files": "**/*",
>     "autoUpload": true,
>     "autoDelete": true
> }
> "syncOption":{
>     "delete": true
> },
> ```
> 
> **!! IMPORTANT !!** Please, be carefull with this configuration as you'll not be able to retrieve you're file(s)/folder(s) if you deleted them on your PC without doing it on purpose. In this case, making backups is the best way.

```json
{
    "name": "NAME",
    "host": "IP",
    "protocol": "sftp",
    "port": 22,
    "secure": true,
    "username": "root",
    "remotePath": "../var/www/html/site",
    "privateKeyPath": "C:\\Users\\ANTON\\.ssh\\id_rsa",
    "uploadOnSave": true,
    "downloadOnOpen": true,
    "useTempFile": false,
    "watcher": {
        "files": "**/*",
        "autoUpload": true,
        "autoDelete": true
    },
    "syncOption": {
        "delete": true
    }
}
```
Checked it helped but caused problem with `"downloadOnOpen": true,`
Gave me an empty file, because I suspect that it caused the file to be asynchronously downloaded when it did not exist or was not yet completely downloaded.
`Ctrl + Z` Helped me

**Comment by NyaPPuu** [comment]
I checked that there is a problem when sending file from a case-insensitive OS to a case-sensitive OS.
like windows → linux.
In Windows, changing the case of the file name does not change uri on the vscode.

I found this solution for get case-sensitive real path in windows : https://stackoverflow.com/a/63760951
And I am testing the code by modifying it as follows:
`src\modules\fileActivityMonitor.ts`
```js
async function handleFileSave(uri: vscode.Uri) {
    ...
    const fspath = await realpathSync.native(uri.fsPath);
    uri = vscode.Uri.file(fspath);
    ...
    await uploadFile(uri);
    ...
}
```
It's working very well right now, and I'll test it a little more.

**Comment by Natizyskunk** [comment]
@NyaPPuu If everything work correctly can you open a new pull-request please ? 🙂

**Comment by NyaPPuu** [comment]
> @NyaPPuu If everything work correctly can you open a new pull-request please ? 🙂

@Natizyskunk 
I don't have a problem.
But I only use Windows, so I have to check if there are any problems with other OSs.

**Comment by ElMarko34** [comment]
Hi.
I tested and this bug still exists


Test procedure (Windows computer to linux ftp target environment)

1) on visual studio: create file: test.php with "hello world" code
2) press ctrl-S -> file is sent to server
3) go to http://www.myserver.com/test.php -> it prompt hello world
4) close test.php
5) rename test.php to Test.php (letter case from t to T)
6) open Test.php and modify "hello world" in "hello world2"
7) press ctrl-S -> console prompt [11-01 06:51:39] [debug] > STOR /test.php 
8) go to http://www.myserver.com/test.php -> it prompt hello world2 
9) go to http://www.myserver.com/Test.php -> file doesn't exist

look at step 7, console prompt update of test.php
step 8: it really updated
step 9: file with upper character is not created

**Comment by NyaPPuu** [comment]
> Hi. I tested and this bug still exists
> 
> Test procedure (Windows computer to linux ftp target environment)
> 
> 1. on visual studio: create file: test.php with "hello world" code
> 2. press ctrl-S -> file is sent to server
> 3. go to http://www.myserver.com/test.php -> it prompt hello world
> 4. close test.php
> 5. rename test.php to Test.php (letter case from t to T)
> 6. open Test.php and modify "hello world" in "hello world2"
> 7. press ctrl-S -> console prompt [11-01 06:51:39] [debug] > STOR /test.php
> 8. go to http://www.myserver.com/test.php -> it prompt hello world2
> 9. go to http://www.myserver.com/Test.php -> file doesn't exist
> 
> look at step 7, console prompt update of test.php step 8: it really updated step 9: file with upper character is not created

@ElMarko34 
Did you test with #249 ?

**Comment by Natizyskunk** [comment]
@ElMarko34,
Some updates ?

**Comment by Natizyskunk** [comment]
@ElMarko34, @microwin7,
A specific `fix-case-rename` version that work with Windows will be added to the Release section but it'll not be added directly to the v1.16.2 untill I'm sure it also work with Mac and Linux.

**Comment by Natizyskunk** [comment]
@ElMarko34, @microwin7,
I've published a new `pre-v1.16.2` pre-release that should fix your problem. You can download it [here](https://github.com/Natizyskunk/vscode-sftp/releases/tag/pre-v1.16.2) or directly from VSCode.
![image](https://user-images.githubusercontent.com/59120131/204685005-69a0b836-cf8f-4187-ae7a-295c5ceddc72.png)
![image](https://user-images.githubusercontent.com/59120131/204685143-2c3be7cb-ecd4-450d-9bc2-90afc982b1ee.png)

**Comment by axgmc** [comment]
Getting the native folder breaks the file saving monitor when working on a symlinked folder!
